src/Covenant/ASG.hs

```diff
@@ -444,6 +538,97 @@ ledgerDestruct ::
+    PvMinor -> Right $ ATyExpr TyInteger -- Integer
+    field -> Left $ TyErrInvalidField ProtocolVersion field
+  ProposalProcedure -> \case
+    PpDeposit -> Right $ ATyLedger (ALedgerNewtype Lovelace) -- Lovelace
+    PpReturnAddr -> Right $ ATyLedger (ALedgerData Credential) -- Credential
+    PpGovernanceAction -> Right $ ATyLedger (ALedgerData GovernanceAction) -- GovernanceAction
+    field -> Left $ TyErrInvalidField ProposalProcedure field
+  TxInInfo -> \case
+    TxInInfoOutRef -> Right $ ATyLedger (ALedgerRecord TxOutRef) --  TxOutRef
+    TxInInfoResolved -> Right $ ATyLedger (ALedgerRecord TxOut) -- TxOut
+    field -> Left $ TyErrInvalidField TxInInfo field
+  TxInfo -> \case
+    TxInfoInputs -> Right $ listOf (ATyLedger (ALedgerRecord TxInInfo)) -- [TxInInfo] **
+    TxInfoReferenceInputs -> Right $ listOf (ATyLedger (ALedgerRecord TxInInfo)) -- [TxInInfo] **
+    TxInfoOutputs -> Right $ listOf (ATyLedger (ALedgerRecord TxOut)) -- [TxOut] **
+    TxInfoFee -> Right $ ATyLedger (ALedgerNewtype Lovelace) -- Lovelace
+    TxInfoMint -> Right $ ATyLedger (ALedgerNewtype MintValue) -- MintValue
+    TxInfoTxCerts -> Right $ listOf (ATyLedger (ALedgerData TxCert)) -- [TxCert] **
+    TxInfoWdrl -> Right $ ATyExpr TyPlutusData -- Map Credential Lovelace **
+    TxInfoValidRange -> Right $ ATyLedger (ALedgerRecord (Interval $ ALedgerNewtype POSIXTime)) -- type POSIXTimeRange = Interval POSIXTime **
+    TxInfoSignatories -> Right $ listOf (ATyLedger (ALedgerNewtype PubKeyHash)) -- [PubKeyHash] **
+    TxInfoRedeemers -> Right $ ATyExpr TyPlutusData -- Map ScriptPurpose Redeemer **
+    TxInfoData -> Right $ ATyExpr TyPlutusData -- Map DatumHash Datum **
+    TxInfoId -> Right $ ATyLedger (ALedgerNewtype TxId) -- TxId
+    TxInfoVotes -> Right $ ATyExpr TyPlutusData -- Map Voter (Map GovernanceActionId Vote) **
+    TxInfoProposalProcedures -> Right $ listOf (ATyLedger (ALedgerRecord ProposalProcedure)) -- [ProposalProcedure] **
+    TxInfoCurrentTreasuryAmount -> Right $ ATyExpr TyPlutusData -- Maybe Lovelace **
+    TxInfoTreasuryDonation -> Right $ ATyExpr TyPlutusData -- Maybe Lovelace **
+    field -> Left $ TyErrInvalidField TxInfo field
+  ScriptContext -> \case
+    ScriptContextTxInfo -> Right $ ATyLedger (ALedgerRecord TxInfo) -- TxInfo
+    ScriptContextRedeemer -> Right $ ATyLedger (ALedgerNewtype Redeemer) -- ScriptInfo
+    ScriptContextScriptInfo -> Right $ ATyLedger (ALedgerData ScriptInfo) -- Credential
+    field -> Left $ TyErrInvalidField ScriptContext field
+  Address -> \case
+    AddressCredential -> Right $ ATyLedger (ALedgerData Credential) -- Credential
+    AddressStakingCredential -> Right $ ATyExpr TyPlutusData -- Maybe StakingCredential **
+    field -> Left $ TyErrInvalidField Address field
+  Interval ty -> \case
+    IvFrom -> Right $ ATyLedger (ALedgerData $ LowerBound ty) -- LowerBound ty **
+    IvTo -> Right $ ATyLedger (ALedgerData $ UpperBound ty) -- UpperBound ty **
```

Comment:
See the comment about parametrising the `Interval` type; same applies in these two cases.